File: TOI-Zero_68/A2/020/020.cpp
#include<bits/stdc++.h>
using namespace std;

#define ll long long

int main(){
	ios::sync_with_stdio(0);
	cin.tie(0);
	int n;
    cin >> n;
    int pass[n+1],visited[n+1],mx=0;
    for(int i=1;i<=n;i++){
    	cin >> pass[i];
    	visited[i]=0;
	}
    for(int i=1;i<=n;i++){
    	if(!visited[i]){
       		int x=i,count=0;
       		while(x!=i || count==0){
       			visited[x]=1;
				count++;
				x=pass[x];
			}
       		mx=max(mx,count);
    	}
	}
    cout << mx;
}

File: Resources/Projects/Archive/2023/Competition/Programming/Contest/TOI-Zero_68/A2/019/TOI-Zero_68/A2/020/020.py
n = int(input())
pass_arr = [0] * (n + 1)
visited = [0] * (n + 1)
mx = 0

for i in range(1, n + 1):
    pass_arr[i] = int(input())
    visited[i] = 0

for i in range(1, n + 1):
    if not visited[i]:
        x = i
        count = 0
        while x != i or count == 0:
            visited[x] = 1
            count += 1
            x = pass_arr[x]
        mx = max(mx, count)

print(mx) 